nvflare/app_common/workflows/base_fedavg.py

```diff
@@ -0,0 +1,176 @@
+
+        - def run(self)
+    """
+
+    def sample_clients(self, min_clients):
+        self._min_clients = min_clients
+
+        clients = self.engine.get_clients()
+        if len(clients) < self._min_clients:
+            self._min_clients = len(clients)
+
+        if self._min_clients < len(clients):
+            random.shuffle(clients)
+            clients = clients[0 : self._min_clients]
+
+        return clients
+
+    @staticmethod
+    def _check_results(results: List[FLModel]):
+        empty_clients = []
+        for _result in results:
+            if not _result.params:
+                empty_clients.append(_result.meta.get("client_name", AppConstants.CLIENT_UNKNOWN))
+
+        if len(empty_clients) > 0:
+            raise ValueError(f"Result from client(s) {empty_clients} is empty!")
+
+    @staticmethod
+    def _aggregate_fn(results: List[FLModel]) -> FLModel:
+        aggregation_helper = WeightedAggregationHelper()
+        for _result in results:
+            aggregation_helper.add(
+                data=_result.params,
+                weight=_result.meta.get(FLMetaKey.NUM_STEPS_CURRENT_ROUND, 1.0),
+                contributor_name=_result.meta.get("client_name", AppConstants.CLIENT_UNKNOWN),
+                contribution_round=_result.meta.get("current_round", None),
+            )
+
+        aggregated_dict = aggregation_helper.get_result()
+
+        aggr_result = FLModel(
+            params=aggregated_dict,
+            params_type=results[0].params_type,
+            meta={"nr_aggregated": len(results), "current_round": results[0].meta["current_round"]},
+        )
+        return aggr_result
+
+    def aggregate(self, results: List[FLModel], aggregate_fn=None) -> FLModel:
+        self.debug("Start aggregation.")
+        self.event(AppEventType.BEFORE_AGGREGATION)
+        self._check_results(results)
+
+        if not aggregate_fn:
+            aggregate_fn = self._aggregate_fn
+
+        self.info(f"aggregating {len(results)} update(s) at round {self._current_round}")
+        try:
+            aggr_result = aggregate_fn(results)
+        except Exception as e:
+            error_msg = f"Exception in aggregate call: {secure_format_exception(e)}"
+            self.exception(error_msg)
+            self.panic(error_msg)
+            return FLModel()
+        self._results = []
+
+        self.fl_ctx.set_prop(AppConstants.AGGREGATION_RESULT, aggr_result, private=True, sticky=False)
+        self.event(AppEventType.AFTER_AGGREGATION)
+        self.debug("End aggregation.")
+
+        return aggr_result
+
+    def update_model(self, aggr_result):
+        self.event(AppEventType.BEFORE_SHAREABLE_TO_LEARNABLE)
+
+        self.model = FLModelUtils.update_model(self.model, aggr_result)
+
+        self.fl_ctx.set_prop(AppConstants.GLOBAL_MODEL, self.model, private=True, sticky=True)
+        self.event(AppEventType.AFTER_SHAREABLE_TO_LEARNABLE)
+
+    def run(self):
+        raise NotImplementedError
```

Comment:
Since `ModelController` already say this method is abstractmethod, we don't need to have it here again.

Reply:
outdated.